#64099: Improve hoisted stylesheet ordering

westonruter opened the pull request:
This primarily ensures that stylesheets related to blocks are inserted right after `wp-block-library` whereas other stylesheets not related to blocks are appended to the end of the `HEAD`. This helps ensure the expected cascade is preserved. If no `wp-block-library` inline style is present, then all styles get appended to the `HEAD` regardless.

When the CSS placeholder comment is removed from the `wp-block-library` inline style, the entire inline style is now removed if there are no styles left (aside from the `sourceURL` comment).

This PR also addresses various deficiencies with the logic in `wp_load_classic_theme_block_styles_on_demand()` and `wp_hoist_late_printed_styles()` after more thorough review.

Lastly, the use of the HTML Tag Processor was significantly improved to leverage `WP_HTML_Text_Replacement`.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/64099

This is a follow-up to https://github.com/WordPress/wordpress-develop/pull/10288 (r61008) and to @dmsnell's [post-merge feedback](https://github.com/WordPress/wordpress-develop/pull/10288#pullrequestreview-3388552834).

# Gemini Review

The changes introduce a new mechanism for hoisting late-printed styles, particularly block-related styles, to the `<head>` section in classic themes. This is a significant improvement for performance and proper CSS cascade.

Here's a detailed review:

**`phpcompat.xml.dist`**

*   **Observation:** A new exclusion rule for `PHPCompatibility.FunctionDeclarations.NewClosure.ThisFoundOutsideClass` has been added for `/src/wp-includes/script-loader.php`.
*   **Review:** This indicates a new closure using `this` outside a class context has been introduced. The comment correctly notes this is a temporary exclusion pending PHPCompatibility v10. This is acceptable as a temporary measure, but it's important to follow up and remove this exclusion once the compatibility issue is resolved.

**`src/wp-includes/script-loader.php`**

*   **Docblock for `wp_print_head_scripts()`:**
    *   **Observation:** The docblock has been updated to provide a clearer explanation of how `wp_hoist_late_printed_styles()` functions, detailing the two sets of styles (block-related and others) and their hoisting.
    *   **Review:** This is a good improvement for code clarity and documentation.
*   **`wp_load_classic_theme_block_styles_on_demand()`:**
    *   **Observation:** Comments have been refined for clarity regarding opt-out mechanisms. A new early exit `if ( ! wp_should_load_separate_core_block_assets() ) { return; }` has been added. The `print_late_styles` filter and its associated comment have been removed.
    *   **Review:** The early exit is a good optimization. The removal of the `print_late_styles` filter suggests a more streamlined approach to handling late styles, which is positive.
*   **`wp_hoist_late_printed_styles()`:**
    *   **Observation:** The `capture_late_styles` closure now captures both `$printed_block_styles` and `$printed_late_styles`. Logic has been added to differentiate and gather block-related styles. `print_late_styles()` is replaced by `wp_styles()->do_footer_items()`. The anonymous `WP_HTML_Tag_Processor` class has new utility methods (`get_span`, `insert_before`, `insert_after`, `remove`). The style insertion logic within the `wp_template_enhancement_output_buffer` filter has been significantly refactored to precisely insert block styles after `wp-block-library-inline-css` and other late styles before `</head>`. It also handles the removal of empty inline style tags.
    *   **Review:** This is the core of the change and appears well-executed. Separating block styles from other late styles is crucial for maintaining the CSS cascade. The use of `WP_HTML_Tag_Processor` for precise HTML manipulation is a robust approach. The logic for removing empty inline style tags (especially those only containing a `sourceURL` comment) is a good detail.
    *   **Nitpick:** The regex `preg_match( ':^/\*# sourceURL=\\S+? \*/$:', trim( $css_text ) )` is quite specific. A brief comment explaining its purpose (to identify an inline style that is effectively empty, containing only a sourceURL comment) would enhance readability.

**`tests/phpunit/tests/template.php`**

*   **`setUp()` and `tearDown()`:**
    *   **Observation:** Several lines related to global `wp_scripts`, `wp_styles`, and `_wp_theme_features` have been removed.
    *   **Review:** This suggests a cleaner test setup and teardown, indicating that these globals are either no longer being directly manipulated or are handled by other mechanisms.
*   **`data_wp_hoist_late_printed_styles()`:**
    *   **Observation:** The data provider now includes `inline_size_limit` and `expected_styles`, and new test cases have been added to cover various scenarios (e.g., min/max inline styles, opting out of separate block styles, disabling block library).
    *   **Review:** This is a substantial improvement in test coverage, ensuring the new logic is thoroughly validated under different configurations.
*   **`test_wp_hoist_late_printed_styles()`:**
    *   **Observation:** The method signature is updated. `switch_theme('default')` is called earlier. Filters for `styles_inline_size_limit` and `wp_get_custom_css` are added. `wp_load_classic_theme_block_styles_on_demand()` and `register_core_block_style_handles()` are explicitly called. A new helper `ensure_style_asset_file_created()` is used to create necessary CSS files for testing. `wp_enqueue_style` now uses `https`. The `wp_hoist_late_printed_styles()` call is conditional. `the_content` filter is applied. Assertions are updated to use the new `$expected_styles` array and a new helper `get_array_snapshot_export()`.
    *   **Review:** These are excellent improvements. The `ensure_style_asset_file_created()` helper makes the tests more robust by not relying on a prior build step. The expanded assertions and the `get_array_snapshot_export()` helper significantly improve the maintainability and clarity of the test suite, especially for snapshot testing. The `@ticket` and `@covers` tags are correctly used.

**Conclusion:**

The changes are well-implemented and significantly improve the handling of late-printed styles in classic themes, particularly for block styles. The accompanying test suite is comprehensive and robust. The temporary `phpcompat.xml.dist` exclusion should be addressed in the future.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

westonruter committed:
Use well-formed HTML in tests

westonruter committed:
Use HTTPS and fix late inline style contents

westonruter committed:
Add separate test case specifically for filtering print_late_styles

westonruter committed:
Add comment for PHPCompatibility issue

westonruter committed:
Use more specific return type for WP_Block_Type_Registry::get_all_registered()

westonruter committed:
Remove PHPCompatibility.FunctionDeclarations.NewClosure.ThisFoundOutsideClass suppression for PHPCS

See https://github.com/WordPress/wordpress-develop/pull/10288#discussion_r2469686284

westonruter committed:
Add a rendered block to the output

westonruter committed:
Add failing test cases for test_wp_hoist_late_printed_styles

westonruter committed:
Exclude PHPCompatibility.FunctionDeclarations.NewClosure.ThisFoundOutsideClass in script-loader.php

dmsnell commented on src/wp-includes/script-loader.php:
```diff
@@ -3685,9 +3685,8 @@ function ( $buffer ) use ( $placeholder, &$printed_late_styles ) {
-					$instance->set_bookmark( 'here' );
-					return $instance->bookmarks['here'];
+					$this->set_bookmark( 'here' );
+					return $this->bookmarks['here'];
```
so much better with the removal of the additional surprising complexity!

dmsnell commented on tests/phpunit/tests/template.php:
```diff
@@ -1109,6 +1216,27 @@ public function assertTemplateHierarchy( $url, array $expected, $message = '' )
 	}
 
+	/**
+	 * Export PHP array as string formatted for pasting into unit test case.
```
a brief example would go a long way. I have a rough idea of what this is supposed to do, which I assume is to aid in debugging test failures, but I really can’t visualize it.

westonruter replied:
Added in 156c6f5cdd991781a0f0593869e957156e91fcb0

dmsnell commented on tests/phpunit/tests/template.php:
```diff
@@ -1109,6 +1216,27 @@ public function assertTemplateHierarchy( $url, array $expected, $message = '' )
+		$export = preg_replace( '/\barray \($/m', 'array(', $export );
+		if ( isset( $snapshot[0] ) ) {
+			$export = preg_replace( '/^(\s+)\d+\s+=>\s+/m', '$1', $export );
+		}
```
we’re trying to remove numeric indices?

westonruter replied:
Yeah, unfortunately `var_export()` includes them, even when sequential and starting from 0.

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

westonruter committed:
Print block styles after wp-block-library and everything else at end of HEAD

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

westonruter committed:
Add example for get_array_snapshot_export

Co-authored-by: Dennis Snell <dmsnell@git.wordpress.org>

westonruter committed:
Prevent enabling should_load_block_assets_on_demand if should_load_separate_core_block_assets was disabled

westonruter committed:
Improve tests

westonruter committed:
Remove obsolete print_late_styles override

westonruter committed:
Update tests

westonruter committed:
Add todo comments

westonruter committed:
Refactor HTML Tag Processor logic

westonruter commented on src/wp-includes/script-loader.php:
```diff
@@ -3695,65 +3722,76 @@ static function () use ( $capture_late_styles ) {
+		function ( $buffer ) use ( $placeholder, &$printed_block_styles, &$printed_late_styles ) {
 
 			// Anonymous subclass of WP_HTML_Tag_Processor which exposes underlying bookmark spans.
 			$processor = new class( $buffer ) extends WP_HTML_Tag_Processor {
```
@dmsnell How does this updated `WP_HTML_Tag_Processor` extension look here? Note the `insert_before` and `insert_after` methods.

westonruter replied:
And now a `review` method as well.

westonruter committed:
Fix missing use of footer in assertion

westonruter committed:
Use correct variable in assertion

westonruter committed:
Remove unhelpful test cases since print_late_styles filter seems to have no effect

westonruter committed:
Set styles_inline_size_limit to unlimited for test

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

westonruter committed:
Opt to set styles_inline_size_limit to zero for test

westonruter committed:
Test that non-empty wp-block-library inline style is not removed

westonruter committed:
Consolidate assertions

westonruter committed:
Fix classic_theme_with_should_load_separate_core_block_assets_opt_out test case

westonruter committed:
Empty out added theme supports before running tests

westonruter committed:
Add core-block-supports-duotone to the list of styles to print after wp-block-library

westonruter committed:
Improve formatting of empty array in export

westonruter committed:
Restore original block type registry after test

westonruter committed:
Fix tests when run as part of enture test suite

westonruter committed:
Revert unrelated WP_Block_Type_Registry change

westonruter committed:
Restore original_ini_config

westonruter committed:
Remove admin-specific style printing logic

westonruter committed:
Carify comments

westonruter committed:
Add test for when styles_inline_size_limit is unlimited

westonruter committed:
Remove obsolete wp_block_styles_not_supported test

westonruter committed:
Remove test for test which is not worthwhile

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

westonruter committed:
Replace todo with explanation of why set_bookmark will never return false

westonruter committed:
Restore comment to _wp_footer_scripts()

westonruter committed:
Restore and update comment for _wp_footer_scripts()

westonruter committed:
Add assertion to ensure wp-block-separator style is registered

westonruter committed:
Debug

westonruter committed:
Undebug

westonruter commented:
This was a big headache. But I believe all the kinks are now worked out!

peterwilsoncc commented:
I've tested this with a [helper plugin](https://gist.github.com/peterwilsoncc/20b1c7cdd22bbb1a660097d3558bc8a4) that includes ~three~ _four_ functions to run and am seeing some regressions between the 6.8 branch and this one when it comes to de registering or modifying the block CSS.

There's various ways a theme may wish to modify the block styles that are now being replaced with Core's styles. 

| Test function | WordPress 6.8 | Feature branch | Result |
| --- | --- | --- | --- |
| `add_dependency_to_block_library` | <img width="1539" height="798" alt="Screenshot 2025-11-07 at 9 35 46 am" src="https://github.com/user-attachments/assets/117fd34b-4437-4b64-a3ab-1a0c67fb992e" /> | <img width="1538" height="804" alt="Screenshot 2025-11-07 at 9 36 38 am" src="https://github.com/user-attachments/assets/f1f2f526-6ee6-4ff3-9f82-ce4ea512c3bf" /> | ✅ |
| `deregister_and_replace_blocks_styles` | <img width="1541" height="807" alt="Screenshot 2025-11-07 at 9 38 31 am" src="https://github.com/user-attachments/assets/ee0b8ca5-b2b4-4560-a936-b29cf80b05a1" /> | <img width="1540" height="767" alt="Screenshot 2025-11-07 at 9 39 02 am" src="https://github.com/user-attachments/assets/5e53d621-7956-47cb-8e51-53f31677acb9" /> | ⚠️ |
| `remove_theme_support_block_styles` | <img width="1540" height="870" alt="Screenshot 2025-11-07 at 9 40 09 am" src="https://github.com/user-attachments/assets/3bb6e306-73b6-48c4-bf76-012d87d7b4c4" /> | <img width="1539" height="820" alt="Screenshot 2025-11-07 at 9 40 54 am" src="https://github.com/user-attachments/assets/4166874e-7588-4265-9938-20f939a430da" /> | ⚠️ |
| `register_custom_block_styles` | <img width="1541" height="693" alt="Screenshot 2025-11-07 at 9 52 40 am" src="https://github.com/user-attachments/assets/7639be12-9bc8-4de7-9c6a-fc57782ff40a" /> | <img width="1541" height="820" alt="Screenshot 2025-11-07 at 9 54 04 am" src="https://github.com/user-attachments/assets/6d76a959-e298-497e-b757-7dd452e53297" /> | ⚠️ |

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

westonruter committed:
Opt for ignore list rather than ack list

westonruter committed:
Only use placeholder comment for ensuring wp-block-library inline style is printed and ensure it is added just in time

westonruter committed:
Improve ensure_style_asset_file_created helper

westonruter commented:
@peterwilsoncc Thank you for this. After evaluating the error scenarios, they all appear to be the same as is described in my [ticket comment](https://core.trac.wordpress.org/ticket/64099#comment:30). Namely, if a site is explicitly expecting `wp-block-library` to be present and _not_ the separate block styles, then they need to add this plugin code:

```php
add_filter( 'should_load_separate_core_block_assets', '__return_false' );
```

This needs to be called out in the dev note for sure.

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into trac-64099-improve-hoisted-stylesheet-ordering

peterwilsoncc committed:
Remove see wrappers of auto-linked function names in dev docs.

westonruter committed:
Add missing static to closure

westonruter committed:
Unimprove ensure_style_asset_file_created helper

Reverts part of 9f816a31ad61f2136780c9107f4a332056a6b599

It turns out the depenency src is empty when a build hasn't been done

westonruter committed:
Update phpcs exclusion after 885bdddd720f45ad13c920f83f1a7a19a3c7acca

peterwilsoncc reviewed (approved):
This looks good for commit.

I've gone through manual testing and reviewed the code in the editor as the diff is a little less clear than reading the code.

As discussed here and in slack, I think we'll need to shout out classic themes replacing styles in the dev note to ensure they are using the correct filters.

Hitting approve as the tests look firmly on their way to passing.

dmsnell commented on src/wp-includes/script-loader.php:
```diff
@@ -3689,65 +3732,99 @@ static function () use ( $capture_late_styles ) {
+				 */
+				private function get_span(): WP_HTML_Span {
+					// Note: This call will never fail according to the usage of this class, given it is always called after ::next_tag() is true.
+					$this->set_bookmark( 'here' );
```
technically it could also fail if there are too many bookmarks already. probably not an issue. in my own code I haven’t been documenting these cases of overlooking the failure given I control the whole flow of execution in the anonymous classes.

dmsnell reviewed (commented):
@westonruter the HTML API usage looks nice. adding those extra methods to the anonymous class does clean up the logic in the part handling the STYLE placeholders.
